File: BW8AGU/VB/Ch05/FirstApp/FirstApp/Common/SuspensionManager.vb
Namespace Common

    ''' <summary>
    ''' SuspensionManager captures global session state to simplify process lifetime management
    ''' for an application.  Note that session state will be automatically cleared under a variety
    ''' of conditions and should only be used to store information that would be convenient to
    ''' carry across sessions, but that should be discarded when an application crashes or is
    ''' upgraded.
    ''' </summary>
    Friend NotInheritable Class SuspensionManager
        Public Shared Async Function SaveAsync() As Task
            Try
                'Save everything in here
            Catch ex As Exception
                Throw New SuspensionManagerException(ex)
            End Try
        End Function

        Public Shared Async Function RestoreAsync() As Task
            Try
                'Restore everything in here
            Catch ex As Exception
                Throw New SuspensionManagerException(ex)
            End Try
        End Function
    End Class

    Public Class SuspensionManagerException
        Inherits Exception
        Public Sub New()
        End Sub

        Public Sub New(ByRef e As Exception)
            MyBase.New("SuspensionManager failed", e)
        End Sub
    End Class
End Namespace

File: BW8AGU/C#/Ch05/FirstApp/Common/SuspensionManager.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Runtime.Serialization;
using System.Text;
using System.Threading.Tasks;
using Windows.Storage;
using Windows.Storage.Streams;
using Windows.UI.Xaml;
using Windows.UI.Xaml.Controls;

namespace FirstApp.Common
{
    internal sealed class SuspensionManager
    {
        public static async Task SaveAsync()
        {
            try
            {
                //Save everything in here
            }
            catch (Exception e)
            {
                throw new SuspensionManagerException(e);
            }
        }

        public static async Task RestoreAsync()
        {
            try
            {
                //Restore everything in here
            }
            catch (Exception e)
            {
                throw new SuspensionManagerException(e);
            }
        }
    }

    public class SuspensionManagerException : Exception
    {
        public SuspensionManagerException()
        {
        }

        public SuspensionManagerException(Exception e)
            : base("SuspensionManager failed", e)
        {

        }
    }
}
